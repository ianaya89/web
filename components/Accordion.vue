<template>
  <div>
    <slot />
  </div>
</template>

<script>
export default {
  name: 'Accordion',

  props: {
    multi: {
      type: Boolean,
      default: false
    }
  },

  data: () => ({
    items: []
  }),

  created: function() {
    this.items = this.$children
    if (!this.multi) {
      this.$on('accordion-item-opened', this.accordionItemOpenedHandler)
    }
  },

  methods: {
    accordionItemOpenedHandler(itemUid) {
      this.items.forEach(item => {
        if (item._uid !== itemUid) {
          item.hide()
        }
      })
    },

    showAll() {
      this.items.forEach(item => {
        item.show()
      })
    },

    hideAll() {
      this.items.forEach(item => {
        item.hide()
      })
    }
  }
}
</script>

<style lang="postcss" scoped>
</style>
