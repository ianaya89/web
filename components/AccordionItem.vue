<template>
  <div :class="[{'border-t' : topBorder}, {'border-b' : bottomBorder}, {'border-l' : leftBorder}, {'border-r' : rightBorder}]" class="accordion border-grey-l3">
    <div class="accordion-header" :class="headerClasses" @click="toggleAccordion">
      <span
        v-if="openerOnLeft"
        :class="['pr-4', currentIconClass]"
        class="opener font-bold"
      />
      <slot name="header" />
      <span
        v-if="!openerOnLeft"
        :class="['px-4', currentIconClass]"
        class="opener font-bold ml-auto"
      />
    </div>
    <transition name="fade-down">
      <div v-if="!closed" :class="[{'hidden' : closed},{'pl-8' : openerOnLeft},bodyClasses]" class="accordion-body">
        <slot />
      </div>
    </transition>
  </div>
</template>

<script>
export default {
  name: 'AccordionItem',

  props: {
    showIcon: {
      type: String,
      default: 'lni-chevron-down'
    },
    hideIcon: {
      type: String,
      default: 'lni-chevron-up'
    },
    headerClasses: {
      type: String,
      default: ''
    },
    bodyClasses: {
      type: String,
      default: ''
    },
    openerOnLeft: {
      type: Boolean,
      default: false
    },
    topBorder: {
      type: Boolean,
      default: false
    },
    bottomBorder: {
      type: Boolean,
      default: true
    },
    leftBorder: {
      type: Boolean,
      default: false
    },
    rightBorder: {
      type: Boolean,
      default: false
    },
    autoOpen: {
      type: Boolean,
      default: false
    }
  },

  data: () => ({
    closed: true
  }),

  computed: {
    currentIconClass() {
      if (this.closed) {
        return this.showIcon
      } else {
        return this.hideIcon
      }
    }
  },

  mounted() {
    this.closed = !this.autoOpen
  },

  methods: {
    toggleAccordion() {
      this.closed = !this.closed
      if (!this.closed) {
        this.$parent.$emit('accordion-item-opened', this._uid)
      }
    },

    hide() {
      this.closed = true
    },

    show() {
      this.closed = false
    }
  }
}
</script>

<style lang="postcss" scoped>
.accordion-header {
  @apply .flex .items-center .select-none .cursor-pointer .relative;
}
.opener {
  @apply text-red;
  &::before {
    @apply .float-right;
  }
}
</style>
