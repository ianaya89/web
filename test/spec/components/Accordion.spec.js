import { mount, shallowMount } from '@vue/test-utils'
import Accordion from '@/components/Accordion'
import AccordionItem from '@/components/AccordionItem'

describe('Accordion', () => {
  it('renders empty accordion', () => {
    const wrapper = shallowMount(Accordion)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders no-mutli accordion', () => {
    const firstItem = {
      render(render) {
        return render(AccordionItem, {
          slots: {
            header: 'First header',
            default: 'Profile fans link'
          }
        })
      }
    }
    const secondItem = {
      render(render) {
        return render(AccordionItem, {
          slots: {
            header: 'Second header',
            default: 'Dashboard'
          }
        })
      }
    }

    const wrapper = mount(Accordion, {
      slots: {
        default: [firstItem, secondItem]
      }
    })
    expect(wrapper.element).toMatchSnapshot()

    const openers = wrapper.findAll('.opener')
    openers.at(0).trigger('click')
    expect(wrapper.element).toMatchSnapshot()

    openers.at(1).trigger('click')
    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders multi accordion', () => {
    const firstItem = {
      render(render) {
        return render(AccordionItem, {
          slots: {
            header: 'First header',
            default: 'Profile fans link'
          }
        })
      }
    }
    const secondItem = {
      render(render) {
        return render(AccordionItem, {
          slots: {
            header: 'Second header',
            default: 'Dashboard'
          }
        })
      }
    }

    const wrapper = mount(Accordion, {
      propsData: {
        multi: true
      },
      slots: {
        default: [firstItem, secondItem]
      }
    })

    expect(wrapper.element).toMatchSnapshot()

    const openers = wrapper.findAll('.opener')
    openers.at(0).trigger('click')
    expect(wrapper.element).toMatchSnapshot()

    openers.at(1).trigger('click')
    expect(wrapper.element).toMatchSnapshot()
  })
})
