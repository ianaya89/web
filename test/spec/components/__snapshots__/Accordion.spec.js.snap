// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion renders empty accordion 1`] = `<div />`;

exports[`Accordion renders multi accordion 1`] = `
<div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`Accordion renders multi accordion 2`] = `
<div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`Accordion renders multi accordion 3`] = `
<div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`Accordion renders no-mutli accordion 1`] = `
<div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`Accordion renders no-mutli accordion 2`] = `
<div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;

exports[`Accordion renders no-mutli accordion 3`] = `
<div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
  <div
    class="accordion border-grey-l3 border-b"
  >
    <div
      class="accordion-header"
    >
      <!---->
        
      <span
        class="opener font-bold ml-auto px-4 lni-chevron-down"
      />
    </div>
     
    <transition-stub
      name="fade-down"
    >
      <!---->
    </transition-stub>
  </div>
</div>
`;
